### PR TITLE
Mean Function

### PR DESCRIPTION
**Pull request by akleeman**
This adds the ability to specify a mean function as well as a covariance function when defining a Gaussian process.

TLDR; using a mean function may be valuable if you want to capture a well observed a priori process as part of the model itself.  For example if you "know" that your data contains a linear trend and are most interested in capturing phenomena which are deviations from that linear trend then using a linear mean could be a good choice.

In most literature Gaussian processes are considered to be zero mean.  To the best of my knowledge this is done because 1) most (all?) mean functions can be turned into an equivalent covariance function 2) most of the computational burden of a GP is in inverting the covariance matrix so the mean function isn't usually the primary concern in literature. 3) you can remove any non-zero mean from the data before fitting a GP.

In practice however there are situations where directly using a mean function is useful.  Take a quadratic function as an example.  A GP with quadratic mean  and arbitrary covariance function, `k`, could be written:
```
gp_with ~ GP(a + b*x + c * x * x, k);
```
This model will have parameters which consist of `a`, `b`, `c` and any params from `k`.  Each of `a`, `b` and `c` can have priors which could also gaussian with associated variances `var_a`, `var_b` and `var_c`.  To determine `a`, `b`, and `c` we would have to use numerical optimization to find the maximum likelihood estimates (or a similar approach)

We could alternatively write a new covariance function and build a new GP:
```
k_quad(x, y) = var_a + var_b * x * y + var_c * x * x * y * y
gp_with_out ~ GP(0, k + k_quad);
```
which should result in very similar predictions, but also has the advantage of having closed form estimates of the quadratic terms along with their uncertainty (which is NOT captured when using the quadratic mean).

So why use the mean directly?  What I've experienced recently is that while packing the quadratic term into a covariance function is much more elegant, it can also lead to some serious numerical round off issues.  Notice the covariance function has terms which are double the order of the equivalent mean function and that the `k_quad` is added to `k`.  If the domain of `x` and `y` get large you can quickly find yourself dealing with extremely large floating point values.

### CODE REVIEW

File: include/albatross/src/covariance_functions/callers.hpp [hunk 1]
```diff
@@ -234,6 +294,28 @@ template <typename SubCaller> struct LinearCombinationCaller {
...
+
+  template <
+      typename MeanFunc, typename X,
+      typename std::enable_if<
+          has_valid_mean_caller<MeanFunc, SubCaller, X>::value, int>::type = 0>
+  static double call(const MeanFunc &mean_func,
+                     const LinearCombination<X> &xs) {
+    double sum = 0.;
+    for (std::size_t i = 0; i < xs.values.size(); ++i) {
+      sum += xs.coefficients[i] * SubCaller::call(mean_func, xs.values[i]);
```
[the remark this answers is not in the document]
Comment (akleeman) [reply include/albatross/src/covariance_functions/callers.hpp]: Yeah, you likely picked up that these "callers" all deal with unwrapping complex feature types.  In this situation we're dealing with a `LinearCombination`, which just means that `xs` here is basically a `std::vector<X>` where each element is scaled by some coefficient.  For example, `LinearCombination` could represent the sum of 2 terms in which case the mean function will be the sum of each of their means.

File: include/albatross/src/covariance_functions/mean_function.hpp [hunk 2]
```diff
@@ -0,0 +1,303 @@
...
+  }
+
+  // Adds the mean to a vector
+  template <typename FeatureType>
+  void add_to(const std::vector<FeatureType> &features,
+              Eigen::VectorXd *target) const {
+    if (std::is_same<Derived, ZeroMean>::value) {
+      return;
+    }
+    const Eigen::VectorXd mean = this->operator()(features);
```
[the remark this answers is not in the document]
Comment (akleeman) [reply include/albatross/src/covariance_functions/mean_function.hpp]: This is just using the method defined on line 69.  We *could* use `call` here instead but the operator has an extra set of type safety checks.  The reason why `operator()` is defined in general is to allow for syntax that looks like:
```
SomeMeanFunction mean_function;
double mean_of_x = mean_function(x);
```

File: include/albatross/src/covariance_functions/callers.hpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (akleeman) [reply include/albatross/src/covariance_functions/callers.hpp]: https://github.com/swift-nav/albatross/pull/191

[the remark this answers is not in the document]
Comment (pgrgich) [reply include/albatross/src/covariance_functions/callers.hpp]: Oh, I see. Yep, sounds good.

File: include/albatross/src/covariance_functions/mean_function.hpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (pgrgich) [reply include/albatross/src/covariance_functions/mean_function.hpp]: Ok, fair enough.
